src/components/Messages/Messages.js

```diff
@@ -23,42 +24,37 @@ class Messages extends Component {
   onListenForMessages = () => {
     this.setState({ loading: true });
 
-    this.props.firebase
+    this.unsubscribe = this.props.firebase
       .messages()
-      .orderByChild('createdAt')
-      .limitToLast(this.state.limit)
-      .on('value', snapshot => {
-        const messageObject = snapshot.val();
-
-        if (messageObject) {
-          const messageList = Object.keys(messageObject).map(key => ({
-            ...messageObject[key],
-            uid: key,
-          }));
-
-          this.setState({
-            messages: messageList,
-            loading: false,
-          });
-        } else {
-          this.setState({ messages: null, loading: false });
-        }
+      .orderBy('createdAt', 'desc')
+      .limit(this.state.limit) // firestore doesn't have limitLast, so we use combination of order desc and limit
+      .onSnapshot(snapshot => {
+        let messageList = [];
+
+        snapshot.forEach(doc =>
+          messageList.push({ ...doc.data(), uid: doc.id }),
+        );
```

Comment:
Do we still need to do the transformation here? I used to do it for the object to array transformation **with the need** to keep track of the `uid`. Isn't the `uid` included by default in the `doc.data()` item? Maybe under a different property name?

Reply:
Suggestion: The `forEach` can be replaced with a `map` array method, because `snapshot` to `messageList` is a 1:1 transformation. One could do:

```
const messageList = snapshot.map(doc => ({ ...doc.data(), uid: doc.id }));
```